Invalid default value for required ZonedDateTime.

##### **Overview of the issue**

Error occurs when field of type ZonedDateTime is marked as required in JDL Studio.

##### **Motivation for or Use Case**

Generator should provide valid default values for every type.

##### **Reproduce the error**

Create entity in JDL studio with required field of ZonedDateTime type.
Import to jhipster project with yo jhipster:import-jdl <file.jh>
Run application with /mvnw command. 

Error:
`2017-05-11 13:42:50.997 ERROR 29125 --- [uler-Executor-1] liquibase                                : classpath:config/liquibase/master.xml: classpath:config/liquibase/changelog/20170510180807_added_entity_Event.xml::20170510180807-1::jhipster: Change Set classpath:config/liquibase/changelog/20170510180807_added_entity_Event.xml::20170510180807-1::jhipster failed.  Error: Invalid default value for 'jhi_stop' [Failed SQL: CREATE TABLE CScheduler.event (id BIGINT AUTO_INCREMENT NOT NULL, name VARCHAR(255) NOT NULL, jhi_start timestamp NOT NULL, jhi_stop timestamp NOT NULL, CONSTRAINT PK_EVENT PRIMARY KEY (id))]
2017-05-11 13:42:51.004 ERROR 29125 --- [uler-Executor-1] i.g.j.c.liquibase.AsyncSpringLiquibase   : Liquibase could not start correctly, your database is NOT ready: Migration failed for change set classpath:config/liquibase/changelog/20170510180807_added_entity_Event.xml::20170510180807-1::jhipster:
     Reason: liquibase.exception.DatabaseException: Invalid default value for 'jhi_stop' [Failed SQL: CREATE TABLE CScheduler.event (id BIGINT AUTO_INCREMENT NOT NULL, name VARCHAR(255) NOT NULL, jhi_start timestamp NOT NULL, jhi_stop timestamp NOT NULL, CONSTRAINT PK_EVENT PRIMARY KEY (id))]

liquibase.exception.MigrationFailedException: Migration failed for change set classpath:config/liquibase/changelog/20170510180807_added_entity_Event.xml::20170510180807-1::jhipster:
     Reason: liquibase.exception.DatabaseException: Invalid default value for 'jhi_stop' [Failed SQL: CREATE TABLE CScheduler.event (id BIGINT AUTO_INCREMENT NOT NULL, name VARCHAR(255) NOT NULL, jhi_start timestamp NOT NULL, jhi_stop timestamp NOT NULL, CONSTRAINT PK_EVENT PRIMARY KEY (id))]
	at liquibase.changelog.ChangeSet.execute(ChangeSet.java:619)
	at liquibase.changelog.visitor.UpdateVisitor.visit(UpdateVisitor.java:51)
	at liquibase.changelog.ChangeLogIterator.run(ChangeLogIterator.java:79)
	at liquibase.Liquibase.update(Liquibase.java:214)
	at liquibase.Liquibase.update(Liquibase.java:192)
	at liquibase.integration.spring.SpringLiquibase.performUpdate(SpringLiquibase.java:431)
	at liquibase.integration.spring.SpringLiquibase.afterPropertiesSet(SpringLiquibase.java:388)
	at io.github.jhipster.config.liquibase.AsyncSpringLiquibase.initDb(AsyncSpringLiquibase.java:82)
	at io.github.jhipster.config.liquibase.AsyncSpringLiquibase.lambda$afterPropertiesSet$0(AsyncSpringLiquibase.java:64)
	at io.github.jhipster.async.ExceptionHandlingAsyncTaskExecutor.lambda$createWrappedRunnable$1(ExceptionHandlingAsyncTaskExecutor.java:66)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: liquibase.exception.DatabaseException: Invalid default value for 'jhi_stop' [Failed SQL: CREATE TABLE CScheduler.event (id BIGINT AUTO_INCREMENT NOT NULL, name VARCHAR(255) NOT NULL, jhi_start timestamp NOT NULL, jhi_stop timestamp NOT NULL, CONSTRAINT PK_EVENT PRIMARY KEY (id))]
	at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:309)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:55)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:113)
	at liquibase.database.AbstractJdbcDatabase.execute(AbstractJdbcDatabase.java:1277)
	at liquibase.database.AbstractJdbcDatabase.executeStatements(AbstractJdbcDatabase.java:1259)
	at liquibase.changelog.ChangeSet.execute(ChangeSet.java:582)
	... 12 common frames omitted
Caused by: com.mysql.jdbc.exceptions.jdbc4.MySQLSyntaxErrorException: Invalid default value for 'jhi_stop'
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at com.mysql.jdbc.Util.handleNewInstance(Util.java:425)
	at com.mysql.jdbc.Util.getInstance(Util.java:408)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:943)
	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3973)
	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3909)
	at com.mysql.jdbc.MysqlIO.sendCommand(MysqlIO.java:2527)
	at com.mysql.jdbc.MysqlIO.sqlQueryDirect(MysqlIO.java:2680)
	at com.mysql.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:2497)
	at com.mysql.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:2455)
	at com.mysql.jdbc.StatementImpl.executeInternal(StatementImpl.java:839)
	at com.mysql.jdbc.StatementImpl.execute(StatementImpl.java:739)
	at com.zaxxer.hikari.pool.ProxyStatement.execute(ProxyStatement.java:95)
	at com.zaxxer.hikari.pool.HikariProxyStatement.execute(HikariProxyStatement.java)
	at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:307)
	... 17 common frames omitted

`



##### **Related issues**

None.

##### **Suggest a Fix**

Provide valid default value for ZonedDateTime.

##### **JHipster Version(s)**

<!--
Which version of JHipster are you using, is it a regression?
-->

##### **JHipster configuration**

`##### **JHipster Version(s)**

```
c-scheduler@0.0.0 /Users/ErnestChechelski/Desktop/jHipsterCScheduler
└── generator-jhipster@4.4.1 

```


##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

```yaml
{
  "generator-jhipster": {
    "promptValues": {
      "packageName": "ernestchechelski.cscheduler",
      "nativeLanguage": "pl"
    },
    "jhipsterVersion": "4.4.1",
    "baseName": "CScheduler",
    "packageName": "ernestchechelski.cscheduler",
    "packageFolder": "ernestchechelski/cscheduler",
    "serverPort": "8080",
    "authenticationType": "jwt",
    "hibernateCache": "ehcache",
    "clusteredHttpSession": false,
    "websocket": "spring-websocket",
    "databaseType": "sql",
    "devDatabaseType": "mysql",
    "prodDatabaseType": "mysql",
    "searchEngine": false,
    "messageBroker": false,
    "serviceDiscoveryType": "eureka",
    "buildTool": "maven",
    "enableSocialSignIn": true,
    "jwtSecretKey": "XDDD",
    "clientFramework": "angular1",
    "useSass": true,
    "clientPackageManager": "yarn",
    "applicationType": "monolith",
    "testFrameworks": [],
    "jhiPrefix": "jhi",
    "enableTranslation": true,
    "nativeLanguage": "pl",
    "languages": [
      "pl",
      "en"
    ]
  }
}
```


##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

Contributor.json

```yaml
{
    "fluentMethods": true,
    "relationships": [
        {
            "relationshipType": "many-to-one",
            "relationshipName": "owner",
            "otherEntityName": "user",
            "otherEntityField": "login"
        }
    ],
    "fields": [
        {
            "fieldName": "alias",
            "fieldType": "String",
            "fieldValidateRules": [
                "required",
                "minlength"
            ],
            "fieldValidateRulesMinlength": 5
        }
    ],
    "changelogDate": "20170510180808",
    "entityTableName": "contributor",
    "dto": "no",
    "pagination": "no",
    "service": "no"
}
```

Event.json

```yaml
{
    "fluentMethods": true,
    "relationships": [],
    "fields": [
        {
            "fieldName": "name",
            "fieldType": "String",
            "fieldValidateRules": [
                "required",
                "minlength"
            ],
            "fieldValidateRulesMinlength": 3
        },
        {
            "fieldName": "start",
            "fieldType": "ZonedDateTime",
            "fieldValidateRules": [
                "required"
            ]
        },
        {
            "fieldName": "stop",
            "fieldType": "ZonedDateTime",
            "fieldValidateRules": [
                "required"
            ]
        }
    ],
    "changelogDate": "20170510180807",
    "entityTableName": "event",
    "dto": "no",
    "pagination": "no",
    "service": "no"
}
```

Group.json

```yaml
{
    "fluentMethods": true,
    "relationships": [],
    "fields": [
        {
            "fieldName": "name",
            "fieldType": "String",
            "fieldValidateRules": [
                "required",
                "minlength"
            ],
            "fieldValidateRulesMinlength": 3
        }
    ],
    "changelogDate": "20170510180809",
    "entityTableName": "group",
    "dto": "no",
    "pagination": "no",
    "service": "no"
}
```

Plan.json

```yaml
{
    "fluentMethods": true,
    "relationships": [],
    "fields": [
        {
            "fieldName": "name",
            "fieldType": "String",
            "fieldValidateRules": [
                "required",
                "minlength"
            ],
            "fieldValidateRulesMinlength": 3
        }
    ],
    "changelogDate": "20170510180810",
    "entityTableName": "plan",
    "dto": "no",
    "pagination": "no",
    "service": "no"
}
```


##### **Browsers and Operating System**

java version "1.8.0_77"
Java(TM) SE Runtime Environment (build 1.8.0_77-b03)
Java HotSpot(TM) 64-Bit Server VM (build 25.77-b03, mixed mode)

git version 2.10.1 (Apple Git-78)

node: v7.10.0

npm: 4.2.0

bower: 1.8.0

gulp:
[13:55:54] CLI version 1.3.0
[13:55:54] Local version 3.9.1

yeoman: 1.8.5

yarn: 0.23.4`


##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

`//User

relationship ManyToOne {
	Contributor{owner(login)} to User
    
}
//Contributors
entity Contributor {
	alias String required minlength(5)
}

//Groups
entity Group {
	name String required minlength(3)
}

//Plans
entity Plan {
	name String required minlength(3)
}

//Events
entity Event {
	name String required minlength(3),
    start ZonedDateTime required,
    stop ZonedDateTime required
}

`

##### **Browsers and Operating System**
OSX El Capitan 10.11.6 (15G1421)

- [ X] Checking this box is mandatory (this is just to show you read everything)

--- a/generators/generator-constants.js
+++ b/generators/generator-constants.js
@@ -20,7 +20,7 @@
 // version of docker images
 const DOCKER_JHIPSTER_REGISTRY = 'jhipster/jhipster-registry:v3.0.0';
 const DOCKER_JAVA_JRE = 'openjdk:8-jre-alpine';
-const DOCKER_MYSQL = 'mysql:5.7.13'; // mysql.5.7.14+ doesn't work well with zoned date time, see https://github.com/jhipster/generator-jhipster/pull/4038
+const DOCKER_MYSQL = 'mysql:5.7.18';
 const DOCKER_MARIADB = 'mariadb:10.1.17';
 const DOCKER_POSTGRESQL = 'postgres:9.6.2';
 const DOCKER_MONGODB = 'mongo:3.2.10';

--- a/generators/server/templates/src/main/docker/_mysql.yml
+++ b/generators/server/templates/src/main/docker/_mysql.yml
@@ -28,4 +28,4 @@ services:
             - MYSQL_DATABASE=<%=baseName.toLowerCase()%>
         ports:
             - 3306:3306
-        command: mysqld --lower_case_table_names=1 --skip-ssl --character_set_server=utf8
+        command: mysqld --lower_case_table_names=1 --skip-ssl --character_set_server=utf8 --explicit_defaults_for_timestamp

--- a/test/templates/compose/01-gateway/src/main/docker/mysql.yml
+++ b/test/templates/compose/01-gateway/src/main/docker/mysql.yml
@@ -1,7 +1,7 @@
 version: '2'
 services:
     jhgate-mysql:
-        image: mysql:5.7.11
+        image: mysql:5.7.18
         # volumes:
         #     - ~/volumes/jhipster/jhgate/mysql/:/var/lib/mysql/
         environment:
@@ -10,4 +10,4 @@ services:
             - MYSQL_DATABASE=jhgate
         ports:
             - 3306:3306
-        command: mysqld --lower_case_table_names=1 --skip-ssl
+        command: mysqld --lower_case_table_names=1 --skip-ssl --character_set_server=utf8 --explicit_defaults_for_timestamp

--- a/test/templates/compose/02-mysql/src/main/docker/mysql.yml
+++ b/test/templates/compose/02-mysql/src/main/docker/mysql.yml
@@ -1,7 +1,7 @@
 version: '2'
 services:
     msmysql-mysql:
-        image: mysql:5.7.11
+        image: mysql:5.7.18
         # volumes:
         #     - ~/volumes/jhipster/msmysql/mysql/:/var/lib/mysql/
         environment:
@@ -10,4 +10,4 @@ services:
             - MYSQL_DATABASE=msmysql
         ports:
             - 3306:3306
-        command: mysqld --lower_case_table_names=1 --skip-ssl
+        command: mysqld --lower_case_table_names=1 --skip-ssl --character_set_server=utf8 --explicit_defaults_for_timestamp

--- a/test/templates/compose/06-uaa/src/main/docker/mysql.yml
+++ b/test/templates/compose/06-uaa/src/main/docker/mysql.yml
@@ -1,7 +1,7 @@
 version: '2'
 services:
     uaaserver-mysql:
-        image: mysql:5.7.11
+        image: mysql:5.7.18
         # volumes:
         #     - ~/volumes/jhipster/uaaserver/mysql/:/var/lib/mysql/
         environment:
@@ -10,4 +10,4 @@ services:
             - MYSQL_DATABASE=uaaserver
         ports:
             - 3306:3306
-        command: mysqld --lower_case_table_names=1 --skip-ssl
+        command: mysqld --lower_case_table_names=1 --skip-ssl --character_set_server=utf8 --explicit_defaults_for_timestamp

--- a/test/templates/compose/08-monolith/src/main/docker/mysql.yml
+++ b/test/templates/compose/08-monolith/src/main/docker/mysql.yml
@@ -1,7 +1,7 @@
 version: '2'
 services:
     samplemysql-mysql:
-        image: mysql:5.7.13
+        image: mysql:5.7.18
         # volumes:
         #     - ~/volumes/jhipster/sampleMysql/mysql/:/var/lib/mysql/
         environment:
@@ -10,4 +10,4 @@ services:
             - MYSQL_DATABASE=samplemysql
         ports:
             - 3306:3306
-        command: mysqld --lower_case_table_names=1 --skip-ssl
+        command: mysqld --lower_case_table_names=1 --skip-ssl --character_set_server=utf8 --explicit_defaults_for_timestamp

--- a/test/templates/compose/09-kafka/src/main/docker/mysql.yml
+++ b/test/templates/compose/09-kafka/src/main/docker/mysql.yml
@@ -1,7 +1,7 @@
 version: '2'
 services:
     samplekafka-mysql:
-        image: mysql:5.7.13
+        image: mysql:5.7.18
         # volumes:
         #     - ~/volumes/jhipster/sampleKafka/mysql/:/var/lib/mysql/
         environment:
@@ -10,4 +10,4 @@ services:
             - MYSQL_DATABASE=samplekafka
         ports:
             - 3306:3306
-        command: mysqld --lower_case_table_names=1 --skip-ssl
+        command: mysqld --lower_case_table_names=1 --skip-ssl --character_set_server=utf8 --explicit_defaults_for_timestamp

